### ISSUE TITLE
neo Update request

### ISSUE THREAD

**Issue by gavin-norman-sociomantic** [sociomantic-tsunami/dhtproto#16]
Many applications need to update records in the DHT by:
1. Get the record's current value,
2. modify it in memory,
3. Put the modified version back to the DHT (overwriting the original value of the record).

The problem is that, if you have multiple instances of such an application running, race conditions can occur when 2 or more are updating the same record at the same time. This has been observed in the real world  and is a known and currently unfixable issue.

It should be possible to keep track of multiple concurrent Update requests on the same record and delay subsequent requests until the initial update has finished:

* When a client starts an Update request, the node allocates an object indicating that the record is locked.
* If another Update starts on the same record, the client is immediately informed that the record is locked.
* The client waits for a short time (10ms?) then restarts the request. (It is important that the client really starts again from the beginning, including re-fetching the list of nodes which are responsible for the record, which could have changed in the meantime, like in the race condition example above.)

Possible usage example, from the client's point of view:
```D
void notifier ( DhtClient.Neo.Update.Notification info, DhtClient.Neo.Update.Args args )
{
  with ( info.Active ) switch ( info.active )
  {
    // may be called multiple times, if other requests change the value
    case received:
      received.update(
        (in void[] record)
        {
          // TODO: update `record` as required
          return updated_record;
        }
      );
    break;

    // TODO: handle other cases...
  }
}

dht.neo.update(channel, key, &notifier);
```

**Comment by gavin-norman-sociomantic** [comment]
A thought about writing a test for this:
1. Start an Update request.
2. When the "here's the record, give me your updated version" notification occurs, spawn a Task which starts a second Update request (for the same record) and then waits (== yield to epoll).
3. The second Update's notification should kick in while the first one's is still suspended, allowing us to test the behaviour of trying to update a record when someone else is already doing so.

**Comment by gavin-norman-sociomantic** [comment]
We'll have to take care that the reading and writing logic works with the carefully designed behaviour of Put and Get to handle online data redistribution.

**Comment by gavin-norman-sociomantic** [comment]
Summary of Put/Get redistribution handling logic:
1. Put requests always write to the newest known node which covers the record hash. (https://github.com/sociomantic-tsunami/dhtproto/blob/neo/src/dhtproto/client/request/internal/Put.d#L151-L160)
2. Get requests try nodes in newest -> oldest order. (https://github.com/sociomantic-tsunami/dhtproto/blob/neo/src/dhtproto/client/request/internal/Get.d#L161-L179)

So the Update request would have to do the same.

**Comment by gavin-norman-sociomantic** [comment]
>It should be possible to keep track of multiple concurrent Update requests on the same record and delay subsequent requests until the initial update has finished.

So there needs to be a fast-lookup data structure in the node which allows data to be temporarily associated with a record (i.e. a hash). When the record is no longer being updated, the associated data is removed. (In that way, we avoid having to permanently store extra metadata with each record. The size of the data structure would max out at the maximum number of records being updated in parallel.)

Of course, there's the risk of a memory leak if these objects are not properly cleaned up as appropriate, after handling Update requests. (Including different error states where an Update does not cleanly finish.)

The simplest way of making the update synchronisation work would be a lock and queue system, as follows:
* When a client starts an Update request, the node allocates an object associated with the record's key. This object contains a queue. If another Update starts on the same record, the request is added to the queue and blocked. When the first Update request finishes (i.e. has written a new value to the record), the next request in the queue is popped and is woken up (given the go-ahead to update the record). When no requests exist in the queue, the object is deallocated.

**Comment by gavin-norman-sociomantic** [comment]
Oof, I think there's a redistribution race condition that could lead to lost updates:
1. Node 1 is responsible for record x.
2. Client A starts an Update on record x and receives the current value from node 1.
3. Client B starts an Update on record x and is queued.
4. Node 2 is now assigned the responsibility for record x. The clients are informed of this change.
5. Client A updates the record and writes the value back to node 2 (which is now the newest known node responsible for record x).
6. Client B's Update is woken up and the value from node 1 sent to the client. *Note:* this record is now out-dated! The new value was stored in node 2, in step 5.
7. Client B updates the record and writes the value back to node 2 (which is now the newest known node responsible for record x). *Note:* the change made to record x by client A, in step 5, is overwritten.

**Comment by gavin-norman-sociomantic** [comment]
Another (actually simpler) approach might be:
* When a client starts an Update request, the node allocates an object indicating that the record is locked. If another Update starts on the same record, the client is immediately informed that the record is locked. The client waits for a short time (10ms?) then restarts the request. (It is important that the client really starts again from the beginning, including re-fetching the list of nodes which are responsible for the record, which could have changed in the meantime, like in the race condition example above.)

**Comment by gavin-norman-sociomantic** [comment]
I _think_ the simple lock/reject proposal would eliminate the risk of race conditions during redistribution. At step 6 (above), client B would start the request again and fetch the correctly updated record from node 2.

**Comment by gavin-norman-sociomantic** [comment]
Another possible scenario is that a record that is being Updated has been moved between nodes by the redistribution. Something like this:
1. Node 1 is responsible for record x.
2. Client A starts an Update on record x and receives the current value from node 1.
3. Node 2 is now assigned the responsibility for record x. The client is informed of this change.
4. Node 1 sends record x to node 2 and removes it from itself.
5. Client A updates the record and writes the value back to node 2 (which is now the newest known node responsible for record x). This would overwrite the value sent from node 1, but this is correct as the record sent by the client is explicitly an update of that record's value.

Seems ok.

**Comment by gavin-norman-sociomantic** [comment]
>Of course, there's the risk of a memory leak if these objects are not properly cleaned up as appropriate, after handling Update requests. (Including different error states where an Update does not cleanly finish.)

For safety, we could consider implementing a cleanup timeout, whereby objects are automatically deallocated if they exist for more than a certain time. (Though actually that would just be a work-around. The real bug would be that a request does not properly deallocate the object.)

**Comment by nemanja-boric-sociomantic** [comment]
>  (Though actually that would just be a work-around. The real bug would be that a request does not properly deallocate the object.)

I agree.

Lock and reject approach sounds reasonable to me, with the user ready to repeat the transaction in case another update is _executing_, which is pretty much standard thing even on relational databases with this isolation level (see serializable isolation level in postgres as an example). One thing that I would like to note at point is that not just `Update` would have to be ready to repeat, but also the `Put` requests for the same record. Of course, we could just rely on the fact that `Put` shouldn't be run on the channel which is managed by `Update`, and leave it as is.


One other thing coming in my mind is making this approach work in reverse: instead of holding new transactions, give them the priority and let know the holders of the previous transactions if they succeeded. So, instead of synchronizing on other transactions executing, serialize on the other transaction succeeding. This would allow for `Put` requests to unconditionally succeed, and it would allow the possible use case of "if there's something newer than me, let it go, don't update anything". Or "Oh, I got this record, but there's a flag that indicates I don't actually want to update this record, oh well". If we want both transactions to succeed, all but one needs to retry anyway.

**Comment by gavin-norman-sociomantic** [comment]
>Of course, we could just rely on the fact that Put shouldn't be run on the channel which is managed by Update, and leave it as is.

This is what I had in mind. Leave this to user good practice. (For now, at least.) Normally there's only one writer per channel, anyway, so I think the risk is low.

>One other thing coming in my mind is making this approach work in reverse

Verbal discussion concluded that this approach would be a bit more complicated to implement (it'd require a hook in the storage engine to "ping" all pending Update requests when a record is modified) and would probably be vulnerable to tricky race conditions during redistribution.

The simpler lock/reject solution seems robust enough for our purposes.

**Comment by gavin-norman-sociomantic** [comment]
Updated issue description with proposed lock/reject scheme.

**Comment by gavin-norman-sociomantic** [comment]
To handle the case where the record was fetched from a different node where it's written back to, the client will have to send an explicit signal to the source node telling it that the request is over (for whatever reason -- including error cases). This will allow the node to unlock the record.

**Comment by gavin-norman-sociomantic** [comment]
Oof this is actually really tricky. In the normal case, this is what happens:
1. Client looks up the responsible node and sends it the Update request.
2. Node locks the record and sends the value to the client.
3. Client modifies the record and sends it back to the node.
4. Node updates the record and unlocks it.

However, during a redistribution, the node to read from and the node to write to can be different. In this situation, the client needs to be in communication with two nodes at once -- the node being read from _and_ the node being written to. The request needs to remain active on the read node until the record has been updated on the write node, so that the client can finally inform the read node that it can unlock the record.

This kind of connection usage is currently not supported by swarm :(

I'll put this request on hold. It looks like it'll be trickier to implement than expected and will require some work in swarm as a prelude.

**Comment by gavin-norman-sociomantic** [comment]
>I'll put this request on hold. It looks like it'll be trickier to implement than expected and will require some work in swarm as a prelude.

See https://github.com/sociomantic-tsunami/swarm/issues/135.

**Comment by gavin-norman-sociomantic** [comment]
What we should do is make sure to log the number of locked records. That way, we'll be able to catch "lock leaks" early.

**Comment by gavin-norman-sociomantic** [comment]
Ooh I just thought of a twist with this request: what to do if the record does not currently exist?

**Comment by gavin-norman-sociomantic** [comment]
An alternative idea that was discussed recently: rather than locking and rejecting records, do as follows:
1. Node: Get the record to be update from storage.
2. Node: Send the record to the client.
3. Client: Hash the original value of the record.
4. Client: Update the record.
5. Client: Send the updated record plus the hash of the original value to the node.
6. Node: When the updated version of the record is received, get the record from storage again.
7. Node: Hash the record value from storage.
8. Node: Compare the hash of the value in storage against the hash sent by the client.
  - If the two hashes match, update the record as requested.
  - If the two hashes do not match, reject the update. (Another client has updated the record in the meantime.)

Pros:
* No need to maintain any extra data structures associated with records. (The risk of this is that if the metadata were to get out of sync, the node could stop functioning properly. Records could become permanently locked, for example.)

Cons:
* Extra CPU hit from hashing record values. We'll have to test this and see if it's problematic in practice.

**Comment by gavin-norman-sociomantic** [comment]
>Extra CPU hit from hashing record values. We'll have to test this and see if it's problematic in practice.

This largely depends on the data in channels that are being Updated. Average record sizes in the order of hundreds of bytes will probably see no impact.

**Comment by gavin-norman-sociomantic** [comment]
#110 updated with a documented compiling (in D1 ;) code, including a usage example. Next steps:
1. Adapt the dhtnode to implement Update.
2. Test.

**Comment by gavin-norman-sociomantic** [comment]
Done in #110.
